Use stack

dkubb opened the pull request:
This branch changes the build process to use stack.

More testing will be required to prevent stack from using a downloaded ghc vs using the system provided ghc. We _always_ want to use the system provided ghc for musl.

dkubb committed:
Change build to use stack instead of cabal

mbj commented on builder/Dockerfile:
```diff
@@ -10,6 +10,6 @@ WORKDIR /src
 RUN until apk add docker; do :; done
 
-RUN ["cabal", "update"]
+RUN ["stack", "upgrade"]
```
Do you consider to nail the stack version to one you explicitly control? So the build is repeatable exactly?

dkubb replied:
I could just use the stack that is pre-installed on the base docker image.

dkubb replied:
@mbj do you think I should remove this altogether? It probably won't break anything existing.

mbj replied:
This would nail the version. Yeah.

dkubb committed:
Move entrypoint and cmd near the bottom of the Dockerfile

* This seems to be convention. I noticed when coming back to this
  project after a few weeks of looking at lots of other Dockerfiles
  "in the wild" I naturally looked at the bottom of the file for these
  directives, and remembered that I put them at the top. I figure this
  may be the case for others so I am moving it lower to make things
  easier for others to understand.

dkubb commented:
@mbj this is ready for review.

I at least figured out how to specify the ghc version that is installed so that if it disagrees with the stack config it will raise an error. I guess this is better than automatically installing a ghc.

mbj commented on example/stack.yaml:
```diff
@@ -0,0 +1,5 @@
+packages:
+- '.'
+extra-deps: []
+resolver: nightly-2015-08-21
```
What about adding a stack version to this file? This would allow you to make sure the stack tool itself is nailed to a specific version. For repeatable builds.

dkubb replied:
> What about adding a stack version to this file?

@mbj I suppose the user could always specify this in their `stack.yaml`. I could provide it as an example and make sure it matches the stack version I am using. I probably won't pin it exactly, but rather I will have it match on the major and minor segments.

mbj commented:
@dkubb I could sucessfully validate this. And only left a tiny flag on the absence of any "fixation" of the stack version in use.

mbj commented:
@dkubb I'm ready to merge. Need write privs to this repo to actually do it.
